Simulation of a PID controlled second order plant.

amcastro-tri opened the pull request:
This PR implements a Simulator test for the `PidControlledSpringMassSystem`. This is an important test since for the first time simulates a series of nested `Diagram<T>`'s.

Similar symbolic wiring of systems is needed for instance in the simulation of the position controlled Kuka arm #3290.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3288)
<!-- Reviewable:end -->

amcastro-tri commented:
Reviewed 3 of 5 files at r1, 1 of 2 files at r2.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@sammy-tri for platform review please.
+@liangfok for feature review please. This is the PR I told you about for which I choose "more interesting" constants for the PID controller.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-:-KSm5UZ4mhz3AlCAUAdj:b-a5nt0k)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 5 files at r1, 1 of 2 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSmH4o8XWVLvhX8nSzG:b-q3usw7) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
> ```cpp
>   // Velocity is computed using AutoDiffScalar.
>   double final_time = 0.2;
>   double x_final, v_final;
> ```

https://google.github.io/styleguide/cppguide.html#Local_Variables says these should be initialized at the point of declaration

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSmK8WFrHwkS0wcgjaS:b-vuraw) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> https://google.github.io/styleguide/cppguide.html#Local_Variables says these should be initialized at the point of declaration
</details>
I believe that reads more like a suggestion for good practice though trying to keep declaration/initialization in the same scope and as close as possible whenever possible. There is the for loop example in which declaration happens outside the for loop scope.
In this case I just don't want to declare a number of auxiliary ugly autodiff variables in the same scope where I need the final restuls, x_final and v_final.  I can add a comment stating this if you think it's appropriate. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSmMbQ2vzSi9S7YGQ32:b-vtnizd) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I believe that reads more like a suggestion for good practice though trying to keep declaration/initialization in the same scope and as close as possible whenever possible. There is the for loop example in which declaration happens outside the for loop scope.
> In this case I just don't want to declare a number of auxiliary ugly autodiff variables in the same scope where I need the final restuls, x_final and v_final.  I can add a comment stating this if you think it's appropriate. 
</details>
Perhaps I was confusing the issue with 
https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#additional-rules regarding class member initialization...

I don't think they need to be initialized to their final values (they're not declared const  after all).  I'd settle for initializing them to double(), as my bigger concern was that we not have variables which contain random garbage from the stack, and double() (default as opposed to an actual value like 0.0) indicates that the real value is coming later.  I'll admit we might be wandering into author's preference territory since it's not likely to be as big of a deal with a local variable vs. a class member.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSqIBrT2ERYsDmQ-UNs:b-uwgdo2) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Perhaps I was confusing the issue with 
> https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#additional-rules regarding class member initialization...
> 
> I don't think they need to be initialized to their final values (they're not declared const  after all).  I'd settle for initializing them to double(), as my bigger concern was that we not have variables which contain random garbage from the stack, and double() (default as opposed to an actual value like 0.0) indicates that the real value is coming later.  I'll admit we might be wandering into author's preference territory since it's not likely to be as big of a deal with a local variable vs. a class member.
</details>
@sammy-tri, I added a comment explaining these variables are properly initialized at the end of this local scope. Hopefully that clarifies?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSqL3DbVzpOBQu0O3D_:b-a03flq) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @sammy-tri, I added a comment explaining these variables are properly initialized at the end of this local scope. Hopefully that clarifies?
</details>
Added value initialization if that helps. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (penjding @liangfok's feature review)

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 238 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSmH4o7NbUe2ovZUXlc:-KSqMLp9KDp4d_86_hXI:b-bk5ip0) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L238)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Added value initialization if that helps. 
</details>
Works for me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-:-KSqMVnqbp6v10cZYRgR:beb90u8)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I completed my first pass feature review.

---

Reviewed 2 of 5 files at r1, 1 of 2 files at r2, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqCgwjbuCmdqktHcZg:-KSqCgwk5soTHIe735LV:bu63yid) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L11)):*
> ```cpp
> 
> #include "drake/systems/framework/examples/spring_mass_system.h"
> #include "drake/systems/framework/examples/controlled_spring_mass_system.h"
> ```

Alpha sort this.

---

*[drake/systems/analysis/test/simulator_test.cc, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqG2YlVJY4rUrqh3iy:-KSqG2YlVJY4rUrqh3iz:bwj8o1o) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L11)):*
> ```cpp
> 
> #include "drake/systems/framework/examples/spring_mass_system.h"
> #include "drake/systems/framework/examples/controlled_spring_mass_system.h"
> ```

BTW: I find it strange for a unit test in`drake-distro/drake/systems/analysis/test/` to instantiate examples in `drake-distro/drake/systems/framework/examples/`. Oh well, let's reorganize things in a future PR after we analyze whether the folder dependencies form a DAG.

---

*[drake/systems/analysis/test/simulator_test.cc, line 173 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqKlIgZQcSfjduEf8j:-KSqKlIh846S2KVycNjY:b-3h9z02) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L173)):*
> ```cpp
> 
> // Tests Simulator for a Diagram system consisting of a tree of systems.
> // For this case the System is a PidControlledSpringMassSystem which is a
> ```

BTW: Consider replacing "For" with "In".

---

*[drake/systems/analysis/test/simulator_test.cc, line 175 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqGZrsO7foQUBmYrVI:-KSqGZrsO7foQUBmYrVJ:b-737w3g) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L175)):*
> ```cpp
> // For this case the System is a PidControlledSpringMassSystem which is a
> // Diagram containing a SpringMassSystem (the plant) and a PidController which
> // in turns is a Diagram composed of primitive systems such as Gain's and
> ```

"turns" should be "turn".

---

*[drake/systems/analysis/test/simulator_test.cc, line 176 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqGdqdmkp00tIcO4Sp:-KSqGdqdmkp00tIcO4Sq:b510tst) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L176)):*
BTW, I believe the "'s" after "Gain" and "Adder" is possessive. Consider replacing:

> primitive systems such as Gain's and Adder's

with:

> primitive elements like Gain and Adder systems.

---

*[drake/systems/analysis/test/simulator_test.cc, line 185 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqLGdcbM63HlkQJzdY:-KSqLGddTnK9X3VywfJa:b-v8gmvb) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L185)):*
> ```cpp
>   const double kMass = 2.0;      // kg
>   // System's open loop frequency.
>   const double wol = sqrt(kSpring / kMass);
> ```

I'm guessing "w" stands for a lowercase Greek omega, "o" stands for "open", and "l" stands for "loop". Is this right? Regardless, can you add a comment explaining what the letters stand for?

---

*[drake/systems/analysis/test/simulator_test.cc, line 195 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqLNQRTIQfO0sMvPr8:-KSqLNQRTIQfO0sMvPr9:b-g5cbde) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L195)):*
> ```cpp
>   const double ki = 0.0;
>   // System's undamped frequency (when kd = 0).
>   const double w0 = sqrt(wol*wol + kp / kMass);
> ```

Missing spaces before and after "*".

---

*[drake/systems/analysis/test/simulator_test.cc, line 208 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqMp_fQgWnAtrZblOs:-KSqMp_gfL6IGyQAuoyZ:b-yg5aqd) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L208)):*
> ```cpp
>   spring_mass.SetDefaultState(simulator.get_mutable_context());
> 
>   // Set initial condition using the Simulator's internal Context.
> ```

BTW: "Set" should be "Sets".

---

*[drake/systems/analysis/test/simulator_test.cc, line 212 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqMz40_EBXa6xQfMfT:-KSqMz40_EBXa6xQfMfU:bsst2qk) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L212)):*
> ```cpp
>   spring_mass.set_velocity(simulator.get_mutable_context(), v0);
> 
>   // Take all the defaults.
> ```

Is the above comment correct?

If it is correct, "Take" should ideally be "Takes".

---

*[drake/systems/analysis/test/simulator_test.cc, line 220 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqN9q0nHp8S_5Fz5Zm:-KSqN9q0nHp8S_5Fz5Zn:b-5mqkbp) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L220)):*
> ```cpp
>   // Computes analytical solution.
>   // 1) Roots of the characteristic equation.
>   complexd lambda1 = -zeta * w0 + w0 * sqrt(complexd(zeta*zeta - 1));
> ```

"zeta*zeta" should be "zeta * zeta".

---

*[drake/systems/analysis/test/simulator_test.cc, line 221 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqNGWpFV8HDTnZ8p6i:-KSqNGWpFV8HDTnZ8p6j:b-5mqkbp) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L221)):*
> ```cpp
>   // 1) Roots of the characteristic equation.
>   complexd lambda1 = -zeta * w0 + w0 * sqrt(complexd(zeta*zeta - 1));
>   complexd lambda2 = -zeta * w0 - w0 * sqrt(complexd(zeta*zeta - 1));
> ```

"zeta*zeta" should be "zeta * zeta".

---

*[drake/systems/analysis/test/simulator_test.cc, line 197 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqNYuJoOLcT-UkZcFL:-KSqNYuK2xuds_YsVMtG:b-ozdtp8) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L197)):*
> ```cpp
>   const double w0 = sqrt(wol*wol + kp / kMass);
>   // Damping ratio (underdamped).
>   const double zeta = 0.5;
> ```

BTW, can you provide a link to a tutorial explaining the meaning of `zeta` and `kd`'s equation?

---

*[drake/systems/analysis/test/simulator_test.cc, line 229 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOM5hMMHu3rSMp5vv:-KSqOM5hMMHu3rSMp5vw:buin6ce) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L229)):*
> ```cpp
>   // The damped frequency corresponds to the absolute value of the imaginary
>   // part of any of the roots.
>   double wd = abs(lambda1.imag());
> ```

BTW, I assume "d" stands for "damped"? Should we create a big table at the top of this test containing all of the variables, what their letters stand for, and their semantics?

This is a relatively small example so a table may not be necessary but setting a gold standard for larger examples to follow may pay off long term.

---

*[drake/systems/analysis/test/simulator_test.cc, line 251 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOuoyWWDnfwBx_FF7:-KSqOuozo34KnKBteHgw:b-z2ze1z) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L251)):*
> ```cpp
>   }
> 
>   // Simulates to final_time.
> ```

BTW, should "to" be "till"?

---

*[drake/systems/analysis/test/simulator_test.cc, line 254 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqP5h8edv4CZUwXjqH:-KSqP5h8edv4CZUwXjqI:brdwr36) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L254)):*
> ```cpp
>   simulator.StepTo(final_time);
> 
>   EXPECT_EQ(simulator.get_num_steps_taken(), 200);
> ```

BTW, can that 200 be computed based on `final_time` and the time step? Will this check be brittle against changes to the integrator?

---

*[drake/systems/analysis/test/simulator_test.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPMPp9cJVRNmNisEH:-KSqPMPp9cJVRNmNisEI:b-l868ti) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L262)):*
> ```cpp
>   EXPECT_EQ(context.get_time(), final_time);  // Should be exact.
> 
>   // Compares with analytical solution (to numerical integration error).
> ```

BTW, should "to" be "modulo"?

---

*[drake/systems/framework/diagram.h, line 240 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPUrHr4Z80YdxGav0:-KSqPUrHr4Z80YdxGav1:bllnr2t) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L240)):*
> ```objc
> 
>   /// Returns a constant reference to the subcontext that corresponds to the
>   /// system @p subsystem.
> ```

BTW, consider deleting "the system" since that's implicit based on the parameter's type and name.

---

*[drake/systems/framework/diagram.h, line 241 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPcQjLLeAeQfD_BcE:-KSqPcQjLLeAeQfD_BcF:b-9posec) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L241)):*
> ```objc
>   /// Returns a constant reference to the subcontext that corresponds to the
>   /// system @p subsystem.
>   /// Classes inheriting from %Diagram need access to this method in order to
> ```

BTW, consider deleting "in order".

---

*[drake/systems/framework/diagram.h, line 242 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPiSOi4VXb5Ikdd5D:-KSqPiSOi4VXb5Ikdd5E:b-rwcs7x) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L242)):*
> ```objc
>   /// system @p subsystem.
>   /// Classes inheriting from %Diagram need access to this method in order to
>   /// pass their constituent subsystem's the apropriate subcontext. Aborts if
> ```

I believe "subsystem's" should be "subsystems".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-:-KSqNTsa8uVJaOFbIT1K:b4e01io)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqCgwjbuCmdqktHcZg:-KSqWV256djoWRIWF1uA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Alpha sort this.
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqG2YlVJY4rUrqh3iy:-KSqWYYshYc2zrqK4Aoo:b-ia0jot) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: I find it strange for a unit test in`drake-distro/drake/systems/analysis/test/` to instantiate examples in `drake-distro/drake/systems/framework/examples/`. Oh well, let's reorganize things in a future PR after we analyze whether the folder dependencies form a DAG.
</details>
BTW, I completely agree with you on this and that is why I decided to just follow the structure of the `SpringMassSystem` test that is already in master. I also hope to solve this DAG thing so we can start placing the correct stuff in `systems/plants`.

---

*[drake/systems/analysis/test/simulator_test.cc, line 173 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqKlIgZQcSfjduEf8j:-KSqXluFZAre1A5pqgcs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L173)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: Consider replacing "For" with "In".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 175 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqGZrsO7foQUBmYrVI:-KSqXpY4NCnF0YRp55Gn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L175)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "turns" should be "turn".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 176 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqGdqdmkp00tIcO4Sp:-KSqXwHESXadJrflhcLe:b-nbcdj1) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L176)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, I believe the "'s" after "Gain" and "Adder" is possessive. Consider replacing:
> 
> > primitive systems such as Gain's and Adder's
> 
> with:
> 
> > primitive elements like Gain and Adder systems.
</details>
I should know this! thanks! Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 185 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqLGdcbM63HlkQJzdY:-KSqY2ekMaBUm_RSYhg6:bd1obe) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L185)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I'm guessing "w" stands for a lowercase Greek omega, "o" stands for "open", and "l" stands for "loop". Is this right? Regardless, can you add a comment explaining what the letters stand for?
</details>
That's the right guess and I knew you wouldn't have any trouble seeing it that's why I didn't comment. I think it'd be a bit of too much of a trivial comment? do you really find it necessary?

---

*[drake/systems/analysis/test/simulator_test.cc, line 195 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqLNQRTIQfO0sMvPr8:-KSqYPLXfGUGzRxVlTDc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L195)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing spaces before and after "*".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 208 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqMp_fQgWnAtrZblOs:-KSqZ6ft0bCzOhW2yV9V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L208)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: "Set" should be "Sets".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 212 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqMz40_EBXa6xQfMfT:-KSqZGnksqqaWniDDDO4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L212)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Is the above comment correct?
> 
> If it is correct, "Take" should ideally be "Takes".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 220 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqN9q0nHp8S_5Fz5Zm:-KSqZLyjh5_jRZrJxEgO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L220)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "zeta*zeta" should be "zeta * zeta".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 221 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqNGWpFV8HDTnZ8p6i:-KSqZMQRmDWrmZdVjRhB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L221)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "zeta*zeta" should be "zeta * zeta".
</details>
Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 197 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqNYuJoOLcT-UkZcFL:-KSqYRPnuNpnqBlw6iIx:b-bbbit1) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L197)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, can you provide a link to a tutorial explaining the meaning of `zeta` and `kd`'s equation?
</details>
BTW, it is from your basic first course on physics in college. You can read Resnick and Halliday's book for instance. It is also very well know for control people. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 229 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOM5hMMHu3rSMp5vv:-KSqZQQwhObDPG1HJsgn:b53hh0l) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L229)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, I assume "d" stands for "damped"? Should we create a big table at the top of this test containing all of the variables, what their letters stand for, and their semantics?
> 
> This is a relatively small example so a table may not be necessary but setting a gold standard for larger examples to follow may pay off long term.
</details>
BTW, I think that'd be too much for this example. I think it is pretty obvious from the context.

---

*[drake/systems/analysis/test/simulator_test.cc, line 251 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOuoyWWDnfwBx_FF7:-KSqZaGnxBJea19nu31S:ba6bqvh) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L251)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, should "to" be "till"?
</details>
BTW, A matter of preferences I think.I like the former since that's how it is usually written in math books.

---

*[drake/systems/analysis/test/simulator_test.cc, line 254 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqP5h8edv4CZUwXjqH:-KSqZmeqwrez0AsUf9_t:b-4wnmrr) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L254)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, can that 200 be computed based on `final_time` and the time step? Will this check be brittle against changes to the integrator?
</details>
BTW, it is a known number for this test. Yes final_time/dt. The same `SpringMassSystem` test above (in master) uses this approach.

---

*[drake/systems/analysis/test/simulator_test.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPMPp9cJVRNmNisEH:-KSq_-p-lUuVCEbhoec4:b-qywwam) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L262)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, should "to" be "modulo"?
</details>
BTW, "to" is the appropriate way to say it, pretty sure.

---

*[drake/systems/framework/diagram.h, line 240 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPUrHr4Z80YdxGav0:-KSqaCRGUmeZeaUN8O5n:b-wg4j0k) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L240)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, consider deleting "the system" since that's implicit based on the parameter's type and name.
</details>
BTW, I am just being consistent with David's style in this file. 

---

*[drake/systems/framework/diagram.h, line 241 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPcQjLLeAeQfD_BcE:-KSqaJWg8SGqjBh1lQIN:b-dgrt54) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L241)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, consider deleting "in order".
</details>
BTW, ditto.

---

*[drake/systems/framework/diagram.h, line 242 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPiSOi4VXb5Ikdd5D:-KSqaQagWh-0WETT2g81:b4ilh7b) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L242)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I believe "subsystem's" should be "subsystems".
</details>
yes, David made the same mistake and I just copied. My bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 2 files at r6.
Review status: 3 of 4 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

liangfok commented:
All of my comments are now non-blocking. Thus <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> .

---

Reviewed 1 of 2 files at r6.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 185 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqLGdcbM63HlkQJzdY:-KSqtmpB3RiIuBbxn15i:brrja3x) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/analysis/test/simulator_test.cc#L185)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> That's the right guess and I knew you wouldn't have any trouble seeing it that's why I didn't comment. I think it'd be a bit of too much of a trivial comment? do you really find it necessary?
</details>
When I first saw the variable, I didn't even know it was an acronym since all of its letters were lower case. Now that I know the naming convention being used throughout this test, I don't think a comment is necessary. I'm afraid, however, that my opinion is tainted by the wisdom gained from experience meaning we should ideally get a fresh pair of eyes on this to see if it's still confusing when it is first encountered.

I will mark this thread as satisfied and leave it to your discretion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 197 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqNYuJoOLcT-UkZcFL:-KSquSSRFzxAUVr64yLm:b-x0m8a1) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L197)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, it is from your basic first course on physics in college. You can read Resnick and Halliday's book for instance. It is also very well know for control people. 
</details>
OK I trust your judgement that this is so basic as to not need documentation.

---

*[drake/systems/analysis/test/simulator_test.cc, line 229 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOM5hMMHu3rSMp5vv:-KSquuhcpZ_F4UxvSjeY:bl00cea) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L229)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, I think that'd be too much for this example. I think it is pretty obvious from the context.
</details>
OK SGTM.

---

*[drake/systems/analysis/test/simulator_test.cc, line 251 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqOuoyWWDnfwBx_FF7:-KSquxd4SJj5WWNNU_p3:b-4ekr2m) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L251)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, A matter of preferences I think.I like the former since that's how it is usually written in math books.
</details>
OK. Thanks.

---

*[drake/systems/analysis/test/simulator_test.cc, line 254 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqP5h8edv4CZUwXjqH:-KSqv09sibWJsHpTR_Vs:bcocps2) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L254)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, it is a known number for this test. Yes final_time/dt. The same `SpringMassSystem` test above (in master) uses this approach.
</details>
OK. As long as it's not brittle against changes to the integrator, I'm fine keeping the literal.

---

*[drake/systems/analysis/test/simulator_test.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPMPp9cJVRNmNisEH:-KSqvAZJXoBqWx0fWess:b2nhktp) ([raw file](https://github.com/robotlocomotion/drake/blob/58b9593c7eccfe6e1c648098925538fecc863951/drake/systems/analysis/test/simulator_test.cc#L262)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, "to" is the appropriate way to say it, pretty sure.
</details>
OK.

---

*[drake/systems/analysis/test/simulator_test.cc, line 172 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqtNoft2NtXbT64Pbb:-KSqtNoft2NtXbT64Pbc:b7224qc) ([raw file](https://github.com/robotlocomotion/drake/blob/3b3b1c2ef2ec5af32bd00d04dd9fb5f0f5d72c9b/drake/systems/analysis/test/simulator_test.cc#L172)):*
> ```cpp
> }
> 
> // Tests Simulator for a Diagram system consisting of a tree of systems.
> ```

BTW, consider changing "for" to "with".

---

*[drake/systems/analysis/test/simulator_test.cc, line 175 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqtX8ACtNrP4m4QmsE:-KSqtX8ACtNrP4m4QmsF:bq5tj3q) ([raw file](https://github.com/robotlocomotion/drake/blob/3b3b1c2ef2ec5af32bd00d04dd9fb5f0f5d72c9b/drake/systems/analysis/test/simulator_test.cc#L175)):*
> ```cpp
> // In this case the System is a PidControlledSpringMassSystem which is a
> // Diagram containing a SpringMassSystem (the plant) and a PidController which
> // in turn is a Diagram composed of primitive systems such as Gain and
> ```

BTW, this sentence now has "systems" twice, which is awkward. Consider replacing the first "systems" with "elements" or "objects".

---

*[drake/systems/framework/diagram.h, line 240 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPUrHr4Z80YdxGav0:-KSqvOreRgXF3ApHRl79:b2nhktp) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L240)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, I am just being consistent with David's style in this file. 
</details>
OK.

---

*[drake/systems/framework/diagram.h, line 241 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPcQjLLeAeQfD_BcE:-KSqvQn-2XpWNCJ-uw0l:b2nhktp) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L241)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> BTW, ditto.
</details>
OK.

---

*[drake/systems/framework/diagram.h, line 242 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqPiSOi4VXb5Ikdd5D:-KSqvU44UBgOFzpiG19U:b8xzyig) ([raw file](https://github.com/robotlocomotion/drake/blob/f5424df4545d30c02b28a721bf0ebc195ad732da/drake/systems/framework/diagram.h#L242)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> yes, David made the same mistake and I just copied. My bad.
</details>
OK. We should ideal fix the original file too, but that should be a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-:-KSqvfZkRSBlywos2Vbm:b28uo18)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 175 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-KSqtX8ACtNrP4m4QmsE:-KSrSFCuDaiReyVUcMiq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3b3b1c2ef2ec5af32bd00d04dd9fb5f0f5d72c9b/drake/systems/analysis/test/simulator_test.cc#L175)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, this sentence now has "systems" twice, which is awkward. Consider replacing the first "systems" with "elements" or "objects".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
-@amcastro-tri 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3288#-:-KSrTT3cKGiVIPB_B2BY:bsv1qg2)*
<!-- Sent from Reviewable.io -->
